File: sampleApp/src/main/java/com/nytimes/android/external/register/sample/SampleActivity.java
package com.nytimes.android.external.register.sample;

import android.animation.ArgbEvaluator;
import android.animation.ValueAnimator;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.design.widget.AppBarLayout;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.util.Pair;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.Switch;
import android.widget.TextView;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.android.billingclient.api.SkuDetails;
import com.android.billingclient.api.SkuDetailsParams;
import com.google.common.collect.ImmutableList;
import com.nytimes.android.external.registerlib.GoogleServiceProvider;
import com.nytimes.android.external.registerlib.GoogleServiceProviderTesting;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

public class SampleActivity extends AppCompatActivity implements CompoundButton.OnCheckedChangeListener {

    static final String TAG = "SampleActivity";
    static final String SKU_IAP = "register.sample.iap";
    static final String SKU_SUB = "register.sample.sub";

    Switch billingApiSwitch;
    ValueAnimator appBarColorAnimator;
    SampleAdapter adapter;
    View emptyView;
    TextView emptyTitle;

    PrefsManager prefsManager;
    GoogleServiceProvider googleServiceProvider;
    CompositeDisposable compositeDisposable;

    PurchasesUpdatedListener purchasesUpdatedListener = (responseCode, purchases) -> checkPurchasesAndSkuDetails();

    BillingClientStateListener billingClientStateListener = new BillingClientStateListener() {
        @Override
        public void onBillingSetupFinished(int responseCode) {
            if (responseCode == BillingClient.BillingResponse.OK) {
                checkPurchasesAndSkuDetails();
            }
        }

        @Override
        public void onBillingServiceDisconnected() {
            //TODO handle disconnection (perhaps reconnection)
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sample);

        initFields();
        initToolbar();
        initEmptyView();
        initRecycler();
        initGoogleServiceProvider();
    }

    @Override
    protected void onPostResume() {
        super.onPostResume();
        checkPurchasesAndSkuDetails();
    }

    @Override
    protected void onDestroy() {
        if (compositeDisposable != null) {
            compositeDisposable.clear();
        }

        if (googleServiceProvider != null) {
            googleServiceProvider.endConnection();
        }

        if (adapter != null) {
            adapter.destroy();
        }
        super.onDestroy();
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        if (billingApiSwitch.equals(buttonView)) {
            prefsManager.setUsingGoogleServiceProvider(isChecked);
            initGoogleServiceProvider();
            checkPurchasesAndSkuDetails();

            if (isChecked) {
                appBarColorAnimator.start();
            } else {
                appBarColorAnimator.reverse();
            }
        }
    }

    void initFields() {
        prefsManager = new PrefsManager(PreferenceManager.getDefaultSharedPreferences(this));
        compositeDisposable = new CompositeDisposable();
    }

    void initToolbar() {
        boolean isUsingTestProvider = prefsManager.isUsingTestGoogleServiceProvider();

        int appBarEnabled = ContextCompat.getColor(this, R.color.colorPrimary);
        int appBarDisabled = ContextCompat.getColor(this, R.color.colorPrimaryAlt);

        AppBarLayout appBarLayout = findViewById(R.id.app_bar_layout);
        appBarLayout.setOnClickListener(v -> billingApiSwitch.toggle());
        appBarLayout.setBackgroundColor(isUsingTestProvider ? appBarEnabled : appBarDisabled);
        appBarLayout.setPadding(appBarLayout.getPaddingLeft(),
                getStatusBarHeight(),
                appBarLayout.getPaddingRight(),
                appBarLayout.getPaddingBottom());

        appBarColorAnimator = ValueAnimator.ofObject(new ArgbEvaluator(), appBarDisabled, appBarEnabled);
        appBarColorAnimator.setCurrentPlayTime(isUsingTestProvider ?
                appBarColorAnimator.getDuration() : 0);
        appBarColorAnimator.addUpdateListener(animator ->
                appBarLayout.setBackgroundColor((int) animator.getAnimatedValue()));

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        billingApiSwitch = findViewById(R.id.app_bar_switch);
        billingApiSwitch.setChecked(isUsingTestProvider);
        billingApiSwitch.setOnCheckedChangeListener(this);
    }

    void initRecycler() {
        adapter = new SampleAdapter(this);
        compositeDisposable.add(adapter.getClickSubject().subscribe(item -> {
            BillingFlowParams params =
                    BillingFlowParams
                            .newBuilder()
                            .setSku(item.getSku())
                            .setType(item.getType())
                            .build();
            googleServiceProvider.launchBillingFlow(this, params);
        }));

        RecyclerView recyclerView = findViewById(R.id.list);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(adapter);
        int dividerSize = getResources().getDimensionPixelSize(R.dimen.recycler_divider_size);
        recyclerView.addItemDecoration(new EmptyItemDecoration(dividerSize));
    }

    void initEmptyView() {
        emptyView = findViewById(R.id.empty_view);
        emptyView.setVisibility(View.GONE);

        emptyTitle = findViewById(R.id.empty_message);
        emptyTitle.setText(prefsManager.isUsingTestGoogleServiceProvider() ?
                R.string.empty_message_register : R.string.empty_message_google);
    }

    void initGoogleServiceProvider() {
        googleServiceProvider =
                GoogleServiceProviderTesting
                        .newBuilder(this)
                        .useTestProvider(prefsManager.isUsingTestGoogleServiceProvider())
                        .setListener(purchasesUpdatedListener)
                        .build();

        googleServiceProvider.startConnection(billingClientStateListener);
    }

    void checkEmptyState() {
        if (adapter.getItemCount() == 0) {
            emptyView.setVisibility(View.VISIBLE);
            emptyTitle.setText(prefsManager.isUsingTestGoogleServiceProvider() ?
                    R.string.empty_message_register : R.string.empty_message_google);
        } else {
            emptyView.setVisibility(View.GONE);
        }
    }

    void checkPurchasesAndSkuDetails() {
        if (googleServiceProvider.isReady()) {
            final SkuDetailsParams iapParams =
                    SkuDetailsParams
                            .newBuilder()
                            .setSkusList(ImmutableList.of(SKU_IAP))
                            .setType(BillingClient.SkuType.INAPP)
                            .build();

            final Observable<List<SkuDetails>> iapSkuDetailsStream =
                    Observable
                            .create(emitter ->
                                    googleServiceProvider
                                            .querySkuDetailsAsync(iapParams, (responseCode, skuDetailsList) -> {
                                                if (skuDetailsList == null) {
                                                    emitter.onNext(new ArrayList<>());
                                                } else {
                                                    emitter.onNext(skuDetailsList);
                                                }
                                                emitter.onComplete();
                                            }));

            final SkuDetailsParams subParams =
                    SkuDetailsParams
                            .newBuilder()
                            .setSkusList(ImmutableList.of(SKU_SUB))
                            .setType(BillingClient.SkuType.SUBS)
                            .build();

            final Observable<List<SkuDetails>> subSkuDetailsStream =
                    Observable
                            .create(emitter ->
                                    googleServiceProvider
                                            .querySkuDetailsAsync(subParams, (responseCode, skuDetailsList) -> {
                                                if (skuDetailsList == null) {
                                                    emitter.onNext(new ArrayList<>());
                                                } else {
                                                    emitter.onNext(skuDetailsList);
                                                }
                                                emitter.onComplete();
                                    }));

            final Observable<List<Purchase>> purchasesResultStream =
                    Observable.
                            fromCallable(() -> {
                                final List<String> skus = ImmutableList.of(SKU_IAP, SKU_SUB);
                                final List<Purchase> purchases = new ArrayList<>();
                                for (String sku : skus) {
                                    Purchase.PurchasesResult result = googleServiceProvider.queryPurchases(sku);
                                    if (result.getResponseCode() == BillingClient.BillingResponse.OK) {
                                        purchases.addAll(result.getPurchasesList());
                                    }
                                }
                                return purchases;
                            });

            compositeDisposable.add(
                    Observable
                            .combineLatest(
                                    iapSkuDetailsStream
                                            .zipWith(
                                                    subSkuDetailsStream,
                                                    (iapSkuDetails, subsSkuDetails) -> {
                                                        final List<SkuDetails> combined = new ArrayList<>();
                                                        combined.addAll(iapSkuDetails);
                                                        combined.addAll(subsSkuDetails);
                                                        return combined;
                                                    }),
                                    purchasesResultStream,
                                    Pair::new
                            )
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(
                                    result -> {
                                        adapter.clear();
                                        for (SkuDetails skuDetails : result.first) {
                                            adapter.addItem(skuDetails);
                                        }
                                        for (Purchase purchase : result.second) {
                                            adapter.addPurchase(purchase);
                                        }
                                        checkEmptyState();
                                    },
                                    throwable -> Log.e(TAG, "Failed to retrieve products and purchases", throwable)
                            )
            );
        }
    }

    public int getStatusBarHeight() {
        int result;
        int resourceId = getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            result = getResources().getDimensionPixelSize(resourceId);
        } else {
            result = getResources().getDimensionPixelSize(R.dimen.status_bar_height);
        }
        return result;
    }
}

File: sampleApp/src/main/java/com/nytimes/android/external/register/sample/SampleActivity.kt
package com.nytimes.android.external.register.sample

import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.os.Bundle
import android.preference.PreferenceManager
import android.util.Log
import android.util.Pair
import android.view.View
import android.widget.CompoundButton
import android.widget.Switch
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.android.billingclient.api.*
import com.google.android.material.appbar.AppBarLayout
import com.nytimes.android.external.registerlib.GoogleServiceProvider
import com.nytimes.android.external.registerlib.GoogleServiceProviderTesting
import com.nytimes.android.external.registerlib.GoogleUtil
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import java.util.*

class SampleActivity : AppCompatActivity(), CompoundButton.OnCheckedChangeListener {

    companion object {
        const val TAG = "SampleActivity"
        const val SKU_IAP = "register.sample.iap"
        const val SKU_SUB = "register.sample.sub"
    }

    private lateinit var adapter: SampleAdapter

    private lateinit var billingApiSwitch: Switch
    private lateinit var appBarColorAnimator: ValueAnimator
    private lateinit var emptyView: View
    private lateinit var emptyTitle: TextView

    private lateinit var prefsManager: PrefsManager
    private lateinit var googleServiceProvider: GoogleServiceProvider
    private var compositeDisposable = CompositeDisposable()

    private val skuMap = mutableMapOf<String, SkuDetails>()

    private var purchasesUpdatedListener = PurchasesUpdatedListener { _, _ -> checkPurchasesAndSkuDetails() }

    private var billingClientStateListener: BillingClientStateListener = object : BillingClientStateListener {
        override fun onBillingSetupFinished(billingResult: BillingResult) {
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                checkPurchasesAndSkuDetails()
            }
        }

        override fun onBillingServiceDisconnected() {
            //TODO handle disconnection case (perhaps reconnection)
        }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sample)

        initFields()
        initToolbar()
        initEmptyView()
        initRecycler()
        initGoogleServiceProvider()
    }

    override fun onDestroy() {
        compositeDisposable.clear()
        googleServiceProvider.endConnection()
        adapter.destroy()

        super.onDestroy()
    }

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        if (billingApiSwitch == buttonView) {
            prefsManager.setUsingGoogleServiceProvider(isChecked)
            initGoogleServiceProvider()
            checkPurchasesAndSkuDetails()

            if (isChecked) {
                appBarColorAnimator.start()
            } else {
                appBarColorAnimator.reverse()
            }
        }
    }

    fun checkPurchasesAndSkuDetails() {
        if (googleServiceProvider.isReady()) {
            val subSkuDetailsStream = getSubSkuDetailsStream()

            val purchasesResultStream = Observable.fromCallable<List<Purchase>> {
                val types = listOf(
                        GoogleUtil.BILLING_TYPE_IAP, GoogleUtil.BILLING_TYPE_SUBSCRIPTION)
                val purchases = ArrayList<Purchase>()
                for (type in types) {
                    val result = googleServiceProvider.queryPurchases(type)
                    if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                        purchases.addAll(result.purchasesList as List<Purchase>)
                    }
                }
                purchases
            }

            getSkuDetails(getIapSkuDetailsStream(subSkuDetailsStream), purchasesResultStream)
        }
    }

    private fun getSubSkuDetailsStream(): Observable<List<SkuDetails>> {
        val subParams = SkuDetailsParams
                .newBuilder()
                .setSkusList(listOf(SKU_SUB))
                .setType(BillingClient.SkuType.SUBS)
                .build()

        return Observable.create { emitter ->
            googleServiceProvider.querySkuDetailsAsync(subParams, SkuDetailsResponseListener { _, skuDetailsList ->
                if (skuDetailsList == null) {
                    emitter.onNext(ArrayList())
                } else {
                    skuDetailsList.forEach {
                        skuMap[it.sku] = it
                    }
                    emitter.onNext(skuDetailsList)
                }
                emitter.onComplete()
            })
        }
    }

    private fun getIapSkuDetailsStream(
            subSkuDetailsStream: Observable<List<SkuDetails>>
    ): Observable<List<SkuDetails>> {
        val iapParams = SkuDetailsParams
                .newBuilder()
                .setSkusList(listOf(SKU_IAP))
                .setType(BillingClient.SkuType.INAPP)
                .build()

        val iapSkuDetailsStream = Observable.create<List<SkuDetails>> { emitter ->
            googleServiceProvider.querySkuDetailsAsync(iapParams, SkuDetailsResponseListener { _, skuDetailsList ->
                if (skuDetailsList == null) {
                    emitter.onNext(ArrayList())
                } else {
                    skuDetailsList.forEach {
                        skuMap[it.sku] = it
                    }
                    emitter.onNext(skuDetailsList)
                }
                emitter.onComplete()
            })
        }

        return iapSkuDetailsStream.flatMap { iapSkuDetails ->
            subSkuDetailsStream.map<List<SkuDetails>> { subsSkuDetails ->
                val combined = ArrayList<SkuDetails>()
                combined.addAll(iapSkuDetails)
                combined.addAll(subsSkuDetails)
                combined
            }
        }
    }

    private fun getSkuDetails(
            iapSkuDetailsStream: Observable<List<SkuDetails>>,
            purchasesResultStream: Observable<List<Purchase>>?
    ) {
        compositeDisposable.add(
                Observable.combineLatest<List<SkuDetails>, List<Purchase>, Pair<List<SkuDetails>, List<Purchase>>>(
                        iapSkuDetailsStream,
                        purchasesResultStream,
                        BiFunction<List<SkuDetails>, List<Purchase>, Pair<List<SkuDetails>, List<Purchase>>> { first, second -> Pair(first, second) }
                )
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                { result ->
                                    adapter.clear()
                                    result.first.forEach { adapter.addItem(it) }
                                    result.second.forEach { adapter.addPurchase(it) }
                                    checkEmptyState()
                                },
                                { throwable -> Log.e(TAG, "Failed to retrieve products and purchases", throwable) }
                        )
        )
    }

    private fun initFields() {
        prefsManager = PrefsManager(PreferenceManager.getDefaultSharedPreferences(this))
        compositeDisposable = CompositeDisposable()
    }

    private fun initToolbar() {
        val isUsingTestProvider = prefsManager.isUsingTestGoogleServiceProvider

        val appBarEnabled = ContextCompat.getColor(this, R.color.colorPrimary)
        val appBarDisabled = ContextCompat.getColor(this, R.color.colorPrimaryAlt)

        val appBarLayout = findViewById<AppBarLayout>(R.id.app_bar_layout)
        appBarLayout.setOnClickListener { billingApiSwitch.toggle() }
        appBarLayout.setBackgroundColor(if (isUsingTestProvider) appBarEnabled else appBarDisabled)
        appBarLayout.setPadding(appBarLayout.paddingLeft,
                getStatusBarHeight(),
                appBarLayout.paddingRight,
                appBarLayout.paddingBottom)

        appBarColorAnimator = ValueAnimator.ofObject(ArgbEvaluator(), appBarDisabled, appBarEnabled)
        appBarColorAnimator.currentPlayTime = if (isUsingTestProvider)
            appBarColorAnimator.duration
        else
            0
        appBarColorAnimator.addUpdateListener { animator -> appBarLayout.setBackgroundColor(animator.animatedValue as Int) }

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        billingApiSwitch = findViewById(R.id.app_bar_switch)
        billingApiSwitch.isChecked = isUsingTestProvider
        billingApiSwitch.setOnCheckedChangeListener(this)
    }

    private fun initEmptyView() {
        emptyView = findViewById(R.id.empty_view)
        emptyView.visibility = View.GONE

        emptyTitle = findViewById(R.id.empty_message)
        emptyTitle.setText(if (prefsManager.isUsingTestGoogleServiceProvider)
            R.string.empty_message_register
        else
            R.string.empty_message_google)
    }

    private fun initRecycler() {
        adapter = SampleAdapter(this)
        compositeDisposable.add(adapter.clickSubject!!.subscribe { item ->
            skuMap[item.sku]?.let { skuDetails ->
                val params = BillingFlowParams
                        .newBuilder()
                        .setSkuDetails(skuDetails)
                        .build()
                googleServiceProvider.launchBillingFlow(this, params)
            }
        })

        val recyclerView = findViewById<RecyclerView>(R.id.list)
        recyclerView.setHasFixedSize(true)
        recyclerView.adapter = adapter
        val dividerSize = resources.getDimensionPixelSize(R.dimen.recycler_divider_size)
        recyclerView.addItemDecoration(EmptyItemDecoration(dividerSize))
    }

    private fun initGoogleServiceProvider() {
        googleServiceProvider = GoogleServiceProviderTesting
                .newBuilder(this)
                .useTestProvider(prefsManager.isUsingTestGoogleServiceProvider)
                .setListener(purchasesUpdatedListener)
               // .enablePendingPurchases() // Not used for subscriptions
                .build()

        googleServiceProvider.startConnection(billingClientStateListener)
    }

    private fun checkEmptyState() {
        if (adapter.itemCount == 0) {
            emptyView.visibility = View.VISIBLE
            emptyTitle.setText(if (prefsManager.isUsingTestGoogleServiceProvider)
                R.string.empty_message_register
            else
                R.string.empty_message_google)
        } else {
            emptyView.visibility = View.GONE
        }
    }

    private fun getStatusBarHeight(): Int {
        val resourceId = resources.getIdentifier("status_bar_height", "dimen", "android")
        return if (resourceId > 0) {
            resources.getDimensionPixelSize(resourceId)
        } else {
            resources.getDimensionPixelSize(R.dimen.status_bar_height)
        }
    }

}
